File: tests/ruby/proton_tests/interop.rb
#!/usr/bin/env ruby

require 'test/unit'
require 'qpid_proton'

if ((RUBY_VERSION.split(".").map {|x| x.to_i}  <=> [1, 9]) < 0)
  require 'pathname'
  class File
    def self.absolute_path(name)
      return Pathname.new(name).realpath
    end
  end
end

class InteropTest < Test::Unit::TestCase
  Data = Qpid::Proton::Data
  Message = Qpid::Proton::Message

  def setup
    @data = Data.new
    @message = Message.new
  end

  # Walk up the directory tree to find the tests directory.
  def get_data(name)
    path = File.absolute_path(__FILE__)
    while path and File.basename(path) != "tests" do path = File.dirname(path); end
    path = File.join(path,"interop")
    raise "Can't find test/interop directory from #{__FILE__}" unless File.directory?(path)
    path = File.join(path,"#{name}.amqp")
    File.open(path, "rb") { |f| f.read }
  end

  # Decode encoded bytes as a Data object
  def decode_data(encoded)
    buffer = encoded
    while buffer.size > 0
      n = @data.decode(buffer)
      buffer = buffer[n..-1]
    end
    @data.rewind
    reencoded = @data.encode
    # Test the round-trip re-encoding gives the same result.
    assert_equal(encoded, reencoded)
  end

  def decode_data_file(name) decode_data(get_data(name)); end

  def decode_message_file(name)
    message = Message.new()
    message.decode(self.get_data(name))
    self.decode_data(message.content)
  end

  def assert_next(type, value)
    assert @data.next
    assert_equal(type, @data.type)
    assert_equal(value, type.get(@data))
  end

  def assert_array_next(expected, header)
    assert_next(Qpid::Proton::ARRAY, expected)
    result = @data.type.get(@data)
    assert_equal(result.proton_array_header, header)
  end

  def test_message
    decode_message_file("message")
    assert_next(Qpid::Proton::STRING, "hello")
    assert !@data.next
  end

  def test_primitives
    decode_data_file("primitives")
    assert_next(Qpid::Proton::BOOL, true)
    assert_next(Qpid::Proton::BOOL, false)
    assert_next(Qpid::Proton::UBYTE, 42)
    assert_next(Qpid::Proton::USHORT, 42)
    assert_next(Qpid::Proton::SHORT, -42)
    assert_next(Qpid::Proton::UINT, 12345)
    assert_next(Qpid::Proton::INT, -12345)
    assert_next(Qpid::Proton::ULONG, 12345)
    assert_next(Qpid::Proton::LONG, -12345)
    assert_next(Qpid::Proton::FLOAT, 0.125)
    assert_next(Qpid::Proton::DOUBLE, 0.125)
    assert !@data.next
  end

  def test_strings
    decode_data_file("strings")
    assert_next(Qpid::Proton::BINARY, "abc\0defg")
    assert_next(Qpid::Proton::STRING, "abcdefg")
    assert_next(Qpid::Proton::SYMBOL, "abcdefg")
    assert_next(Qpid::Proton::BINARY, "")
    assert_next(Qpid::Proton::STRING, "")
    assert_next(Qpid::Proton::SYMBOL, "")
    assert !@data.next
  end

  def test_described
    decode_data_file("described")
    assert_next(Qpid::Proton::DESCRIBED, Qpid::Proton::Described.new("foo-descriptor", "foo-value"))
    assert(@data.described?)
    assert_next(Qpid::Proton::DESCRIBED, Qpid::Proton::Described.new(12, 13))
    assert(@data.described?)
    assert !@data.next
  end

  def test_described_array
    decode_data_file("described_array")
    assert_array_next((0...10).to_a,
                       Qpid::Proton::ArrayHeader.new(Qpid::Proton::INT,
                                                     "int-array"))
  end

  def test_arrays
    decode_data_file("arrays")
    assert_array_next((0...100).to_a,
                      Qpid::Proton::ArrayHeader.new(Qpid::Proton::INT))
    assert_array_next(["a", "b", "c"],
                      Qpid::Proton::ArrayHeader.new(Qpid::Proton::STRING))
    assert_array_next([],
                      Qpid::Proton::ArrayHeader.new(Qpid::Proton::INT))
    assert !@data.next
  end

  def test_lists
    decode_data_file("lists")
    assert_next(Qpid::Proton::LIST, [32, "foo", true])
    assert_next(Qpid::Proton::LIST, [])
    assert !@data.next
  end

  def test_maps
    decode_data_file("maps")
    assert_next(Qpid::Proton::MAP, {"one" => 1, "two" => 2, "three" => 3 })
    assert_next(Qpid::Proton::MAP, {1 => "one", 2 => "two", 3 => "three"})
    assert_next(Qpid::Proton::MAP, {})
    assert !@data.next
  end
end

File: python/tests/proton_tests/interop.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from __future__ import absolute_import

import os

from proton import *

from . import common


def find_test_interop_dir():
    """Find the common tests directory relative to this script"""
    from os.path import dirname, join, abspath, isdir
    f = dirname(dirname(dirname(dirname(abspath(__file__)))))
    f = join(f, "tests", "interop")
    if not isdir(f):
        raise Exception("Cannot find tests/interop directory from "+__file__)
    return f

test_interop_dir=find_test_interop_dir()

class InteropTest(common.Test):

    def setUp(self):
        self.data = Data()
        self.message = Message()

    def tearDown(self):
        self.data = None

    def get_data(self, name):
        filename = os.path.join(test_interop_dir, name+".amqp")
        f = open(filename,"rb")
        try: return f.read()
        finally: f.close()

    def decode_data(self, encoded):
        buffer = encoded
        while buffer:
            n = self.data.decode(buffer)
            buffer = buffer[n:]
        self.data.rewind()

    def decode_data_file(self, name):
        encoded = self.get_data(name)
        self.decode_data(encoded)
        encoded_size = self.data.encoded_size()
        # Re-encode and verify pre-computed and actual encoded size match.
        reencoded = self.data.encode()
        assert encoded_size == len(reencoded), "%d != %d" % (encoded_size, len(reencoded))
        # verify round trip bytes
        assert reencoded == encoded, "Value mismatch: %s != %s" % (reencoded, encoded)

    def decode_message_file(self, name):
        self.message.decode(self.get_data(name))
        body = self.message.body
        if str(type(body)) == "<type 'org.apache.qpid.proton.amqp.Binary'>":
            body = body.array.tostring()
        self.decode_data(body)

    def assert_next(self, type, value):
        next_type = self.data.next()
        assert next_type == type, "Type mismatch: %s != %s"%(
            Data.type_names[next_type], Data.type_names[type])
        next_value = self.data.get_object()
        assert next_value == value, "Value mismatch: %s != %s"%(next_value, value)

    def test_message(self):
        self.decode_message_file("message")
        self.assert_next(Data.STRING, "hello")
        assert self.data.next() is None

    def test_primitives(self):
        self.decode_data_file("primitives")
        self.assert_next(Data.BOOL, True)
        self.assert_next(Data.BOOL, False)
        self.assert_next(Data.UBYTE, 42)
        self.assert_next(Data.USHORT, 42)
        self.assert_next(Data.SHORT, -42)
        self.assert_next(Data.UINT, 12345)
        self.assert_next(Data.INT, -12345)
        self.assert_next(Data.ULONG, 12345)
        self.assert_next(Data.LONG, -12345)
        self.assert_next(Data.FLOAT, 0.125)
        self.assert_next(Data.DOUBLE, 0.125)
        assert self.data.next() is None

    def test_strings(self):
        self.decode_data_file("strings")
        self.assert_next(Data.BINARY, b"abc\0defg")
        self.assert_next(Data.STRING, "abcdefg")
        self.assert_next(Data.SYMBOL, "abcdefg")
        self.assert_next(Data.BINARY, b"")
        self.assert_next(Data.STRING, "")
        self.assert_next(Data.SYMBOL, "")
        assert self.data.next() is None

    def test_described(self):
        self.decode_data_file("described")
        self.assert_next(Data.DESCRIBED, Described("foo-descriptor", "foo-value"))
        self.data.exit()

        assert self.data.next() == Data.DESCRIBED
        self.data.enter()
        self.assert_next(Data.INT, 12)
        self.assert_next(Data.INT, 13)
        self.data.exit()

        assert self.data.next() is None

    def test_described_array(self):
        self.decode_data_file("described_array")
        self.assert_next(Data.ARRAY, Array("int-array", Data.INT, *range(0,10)))

    def test_arrays(self):
        self.decode_data_file("arrays")
        self.assert_next(Data.ARRAY, Array(UNDESCRIBED, Data.INT, *range(0,100)))
        self.assert_next(Data.ARRAY, Array(UNDESCRIBED, Data.STRING, *["a", "b", "c"]))
        self.assert_next(Data.ARRAY, Array(UNDESCRIBED, Data.INT))
        assert self.data.next() is None

    def test_lists(self):
        self.decode_data_file("lists")
        self.assert_next(Data.LIST, [32, "foo", True])
        self.assert_next(Data.LIST, [])
        assert self.data.next() is None

    def test_maps(self):
        self.decode_data_file("maps")
        self.assert_next(Data.MAP, {"one":1, "two":2, "three":3 })
        self.assert_next(Data.MAP, {1:"one", 2:"two", 3:"three"})
        self.assert_next(Data.MAP, {})
        assert self.data.next() is None
